docs/design/2018-10-19-table-partition.md

```diff
@@ -0,0 +1,147 @@
+Now that table partition is adding, we have to consider the following cases:
+
+1. New TiDB runs on the old cluster data
+2. Old TiDB runs on the new cluster data
+3. Upgrading phases during part of the features is implemented
+
+A new `PartitionInfo.Enable` flag is introduced to the `TableInfo`, and it's persisted.
+New TiDB runs on the old cluster will check this flag and find `PartitionInfo.Enable` is false, so it handles the table as before.
+Old TiDB will not be able to run with partitioned table data, so **the upgrading is incompatible**. (Note that if no partitioned table is created with the new TiDB, i.e, the feature is not used, the cluster is still able to degrade)
+
+If partition by range is implemented while partition by hash is not, TiDB will not set `PartitionInfo.Enable` to true for `create table ... partition by hash ...`, but it will set do that for `create table ... partition by range ...`.
+
+In a word, only when the persisted `PartitionInfo.Enable` is true, and the code could handle partitioned table, will the table partition feature be supported.
+
+## Implementation
+
+### How the partitioned table is stored
+
+There are two levels of mapping in TiDB: SQL data to a logical key range, logical key range to the physical storage.
+When TiDB maps table data to key-value storage, it first encodes `table id + row id` as key, row data as value. Then the logical key range is split into regions and stored into TiKV.
+
+Table partition works on the first level of mapping, partition ID will be made equivalent of the table ID. A partitioned table row uses `partition id + row id` as encoded key.
+
+Table ID and partition IDs relationship should be maintained in the `TableInfo`.
+
+When a new row is inserted into a partitioned table, if the row doesn't  belong to any partition of the table, the operation fails.
+
+Handling `NULL` values refers to MySQL document https://dev.mysql.com/doc/refman/8.0/en/partitioning-handling-nulls.html
+
+### How to read from the partitioned table
+
+```
+create table t (id int) partition by range (id)
+(partition p1 values less than (10),
+partition p2 values less than (20),
+partition p3 values less than (30))
+```
+
+This SQL query
+
+```
+select * from t 
+```
+
+is equal to
+
+```
+select * from (union all
+select * from p1 where id < 10
+select * from p2 where id < 20
+select * from p3 where id < 30)
+```
+
+During logical optimize phase, the `DataSource` plan is translated into `UnionAll`, then each partition generates its own `TableReader` in the physical optimize phase.
+
+The drawbacks of this implementation are:
+
+* If the table has many partitions, there will be many readers, then `explain` result is not friendly to user
+* `UnionAll` executor can't keep order, so if some executor need ordered result such as `IndexReader`, an extra `Sort` executor is needed
```

Comment:
The `UnionAll` executor cannot guarantee that the output results are in order, so if some executor needs an ordered result such as `IndexReader`, an extra `Sort` executor is needed